torch/_inductor/scheduler.py

```diff
@@ -861,6 +965,154 @@ def check_all_pairs(nodes):
+        # in different orders. So, we inspect the MemoryDep's indexing
+        # expressions to find a reordering that makes them equal.
+        #
+        # Note that since nodes can have multiple buffers in common, this may
+        # produce duplicate reorderings so we build the full set of reorderings
+        # first to eliminate duplicates.
+        reorder_opportunities: Dict[
+            Tuple[BaseSchedulerNode, BaseSchedulerNode], Set[Tuple[int, ...]]
+        ] = collections.defaultdict(set)
+
+        for buf_name, node_grouping in buffer_names_grouping.items():
+            for node1, node2 in itertools.combinations(node_grouping, 2):
+                if (node1, node2) in ignore:
+                    continue
+
+                if node1.is_reduction() and node2.is_reduction():
+                    continue
+
+                if any(
+                    not isinstance(n, (FusedSchedulerNode, SchedulerNode))
+                    for n in itertools.chain(node1.get_nodes(), node2.get_nodes())
+                ):
+                    continue
+
+                def node_ndim(node):
+                    pointwise, reduction = node.get_ranges()
+                    return len(pointwise) + len(reduction)
+
+                ndim = max(node_ndim(n) for n in node1.get_nodes())
+
+                def filter_in(dep):
+                    return (
+                        isinstance(dep, MemoryDep)
+                        and dep.name == buf_name
+                        and len(dep.size) == ndim
+                    )
+
+                node1_buf_deps = [
+                    dep
+                    for dep in (node1.read_writes.reads | node1.read_writes.writes)
+                    if filter_in(dep)
+                ]
+                node2_buf_deps = [
+                    dep
+                    for dep in (node2.read_writes.reads | node2.read_writes.writes)
+                    if filter_in(dep)
+                ]
+
+                for dep1, dep2 in itertools.product(node1_buf_deps, node2_buf_deps):
+                    if dep1 == dep2:
+                        continue
+
+                    reorder = self.pick_dependency_reorder(dep1, dep2)
+                    if reorder is not None:
+                        reorder_opportunities[(node1, node2)].add(tuple(reorder))
+
+        # Step 2: Construct new nodes with the reordering and comprehesively
+        # test if they can be fused with self.can_fuse
+        for (node1, node2), orders in reorder_opportunities.items():
+            for order in orders:
+                # Prefer to reorder pointwise ops, or node2 if both match
+                reorder_node1 = node2.is_reduction() and not node1.is_reduction()
+                if reorder_node1:
+                    assert isinstance(node1, (SchedulerNode, FusedSchedulerNode))
+                    new_node = node1.reorder_loop(ir.inverse_reorder(order), len(order))
```

Comment:
Should we cache this on the node?  Since it requires retracing the body it might be slow.